api/web3server.go

```diff
@@ -431,19 +431,48 @@ func (svr *Web3Server) estimateGas(in interface{}) (interface{}, error) {
 	}
 
 	var estimatedGas uint64
-	if isContract, _ := svr.isContractAddr(to); isContract {
+	actType, err := svr.actionTypeByAddr(to)
+	if err != nil {
+		return nil, err
+	}
+	switch actType {
+	case _executionAction:
 		exec, _ := action.NewExecution(to, 0, value, gasLimit, big.NewInt(0), data)
 		estimatedGas, err = svr.coreService.EstimateExecutionGasConsumption(context.Background(), exec, from)
-		if err != nil {
-			return nil, err
-		}
-	} else {
-		// TODO: support staking actions
-		estimatedGas, err = svr.coreService.CalculateGasConsumption(action.TransferBaseIntrinsicGas, action.TransferPayloadGas, uint64(len(data)))
-		if err != nil {
-			return nil, err
+	case _transferAction:
+		estimatedGas, err = svr.coreService.EstimateGasV2(&action.Transfer{}, uint64(len(data)))
+	case _stakingAction:
+		if len(data) <= 4 {
+			return nil, errInvalidFormat
+		}
+		var methodID [4]byte
+		copy(methodID[:], data[:4])
+		switch methodID {
+		case action.CreateStakeMethodID:
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.CreateStake{}, uint64(len(data)))
+		case action.DepositToStakeMethodID:
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.DepositToStake{}, uint64(len(data)))
+		case action.ChangeCandidateMethodID:
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.ChangeCandidate{}, uint64(len(data)))
+		case action.UnstakeMethodID:
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.Unstake{}, uint64(len(data)))
+		case action.WithdrawStakeMethodID:
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.WithdrawStake{}, uint64(len(data)))
+		case action.RestakeMethodID:
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.Restake{}, uint64(len(data)))
+		case action.TransferStakeMethodID:
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.TransferStake{}, uint64(len(data)))
+		case action.CandidateRegisterMethodID:
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.CandidateRegister{}, uint64(len(data)))
+		case action.CandidateUpdateMethodID:
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.CandidateUpdate{}, 0)
+		default:
+			return nil, errInvalidFormat
 		}
 	}
+	if err != nil {
```

Comment:
estimatedGas, err = svr.coreService.EstimateGasForNonExecution(actType, len(data))

Reply:
yes, what i mean is `actType = "stakeCreate", "stakeUnstake"` for each different staking tx